Path Unit: Fix  ENOSPC issue during regression testing of path unit.

bpunnuru opened the pull request:
we were using path unit to monitor a directory for minidumps.
           We made a process to constantly producing a minidump in monotored directory.
           After a while we keep seeing error ENOSPC, while adding a watch.

           Fixed by adding inotify_rm_watch, while unwatching path.
           Also along with fixed issue where  we are not returning
           correct watch descriptor from inotify_add_watch_and_warn.

Signed-off-by: Balaji Punnuru <balaji_punnuru@cable.comcast.com>

poettering commented on src/basic/fs-util.c:
```diff
@@ -705,15 +705,17 @@ int inotify_add_watch_fd(int fd, int what, uint32_t mask) {
 }
 
 int inotify_add_watch_and_warn(int fd, const char *pathname, uint32_t mask) {
+        int r = -1 ;
```
no need to initialize this, we'll overrid in the next line anyway.

also, there's a spurious space before the ;

bpunnuru replied:
incorporated comments.

poettering commented on src/core/path.c:
```diff
@@ -144,6 +144,7 @@ void path_spec_unwatch(PathSpec *s) {
         assert(s);
 
         s->event_source = sd_event_source_unref(s->event_source);
+        inotify_rm_watch(s->inotify_fd,s->primary_wd);
```
this appears unnecessary, we close the inotify ID right after, which which should remove all descriptors implicitly for us.

can you explain?

bpunnuru replied:
Your statement  about closing file descriptor cleaning up of watches is correct. But what about if we see failures in closing? Kernel will still maintain the watch counts.

We are running systems in our arm based devices, 

The same code was working in 3.x kernel. But in 4.x kernels there are som issues around closing the file descriptors, due to which the watch count in the kernel space does not decrement.  Hence found reasonable to add code to inotify_rm_watch, which will reduce the counter, even if there are issues in closing the file descriptor.

poettering replied:
we generally try to avoid working around kernel bugs. If the watch count doesn't decrease even if the responsible inotify fd is fully closed, then that's a kernel bug, and really should be fixed there.

poettering commented:
Can you explain the inotify_rm_watch (see earlier comment) the closing of the inotify should implicitly destroy all wds...

bpunnuru commented:
Your statement about closing file descriptor cleaning up of watches is correct. But what about if we see failures in closing? Kernel will still maintain the watch counts.

We are running systems in our arm based devices,

The same code was working in 3.x kernel. But in 4.x kernels there are som issues around closing the file descriptors, due to which the watch count in the kernel space does not decrement. Hence found reasonable to add code to inotify_rm_watch, which will reduce the counter, even if there are issues in closing the file descriptor.

poettering commented:
As mentioned I am not keen on working around such kernel bugs, let's let that be fixed by the kernel people. The other part of your patch does fix a major bug in systemd however. Any chance you can rework it, and drop the watch removal, keeping the rest? Would love to merge it then!

poettering commented:
I have split the relevant commit now out, into #15514 and cleaned it up a bit. I made sure to keep you as author in, so that you are credited for this in the commit logs.

Hope it's OK to close this here hence, in favour of #15514

bpunnuru commented:
It is ok with me. Thanks for taking in the part of fix.

Regards,
Balaji Punnuru.

On Tue, Apr 21, 2020 at 11:19 AM Lennart Poettering <
notifications@github.com> wrote:

> I have split the relevant commit now out, into #15514
> <https://github.com/systemd/systemd/pull/15514> and cleaned it up a bit.
> I made sure to keep you as author in, so that you are credited for this in
> the commit logs.
>
> Hope it's OK to close this here hence, in favour of #15514
> <https://github.com/systemd/systemd/pull/15514>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/systemd/systemd/pull/15381#issuecomment-617246810>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALGMQDFBBAU4B6SIQ3TZ5LTRNW2Q7ANCNFSM4ME3UK2A>
> .
>

poettering commented:
thanks for tracking it down!
